src/gpio.rs

```diff
@@ -2,17 +2,29 @@
 //!
 //! Interface for the input/output pins.
 //!
-//! `Gpio1` through `Gpio39` represent the *physical* pins of the ESP chip.
+//! `Gpio1` through `GpioNN` represent the pin peripherals of the ESP chip. You
+//! may think of pin peripherals as *physical* pins.
 //!
-//! *Logical* pins (compatible with `embedded_hal::digital::InputPin`/`OutputPin`)
-//! are implemented through `PinDriver`.
+//! Pin drivers are implemented through `PinDriver`. You may think of pin drivers as
+//! *logical* I/O pins. They implement the [`embedded_hal::digital::InputPin`](https://docs.rs/embedded-hal/latest/embedded_hal/digital/trait.InputPin.html)/[`OutputPin`](https://docs.rs/embedded-hal/latest/embedded_hal/digital/trait.OutputPin.html)
+//! traits, as well as their [`embedded_hal_async`](https://docs.rs/embedded-hal-async/latest/embedded_hal_async/) counterparts.
 //!
-//! The ESP architecture has a I/O multiplexer, which means that (almost) any
-//! physical pin can be used for any logical function (i.e. GPIO, I2C, SPI, ADC, etc).
-//! Even though it's possible to use a pin for several functions at once, this
-//! should be avoided. This is particularly important with the pins used for the
+//! The ESP architecture has a [I/O multiplexer](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#iomuxgpio),
+//! which means that (almost) any physical pin can be used for any logical
+//! function (i.e. GPIO, I2C, SPI, ADC, etc). Even though it's possible to
+//! use a pin for several functions at once, this should be avoided. In
+//! practice, `esp-idf-hal` should prevent most instances of pin reuse.
+//!
+//! If you *really* need to mux I/O pins, you might need to drop any function
```

Comment:
I think this part would be confusing for people as it doesn't cleary describe what you mean here.

 For example most of our logical drivers accept also pin references. E.g you can give a PinDriver a &mut gpio0 and if you drop the PinDriver you still have access to gpio0 instance. No need for fancy Mutex etc.
 
 ```rust
     let mut bla = per.pins.gpio0.downgrade_input();

    let driver = PinDriver::input(&mut bla);
    drop(driver);
    let driver2 = PinDriver::input(&mut bla);
    ```